### PR TITLE
Dev dagger

### PR DESCRIPTION
**Pull request by vabdollahi**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -108,6 +108,18 @@ Here are a few examples:
   $ python -m main +experiment=simple_dqn/connect_four +run.hill_training_trials_ratio=0.05
   ```
 
+- Launch a [DAGGER](https://arxiv.org/abs/1011.0686) imitation learning algorithm by first training the expert using the simple_a2c method
```
Comment (cloderic) [inline README.md:111]: Can you create a dedicated "page" for DAGGER (in docs/results) that have this exact content and maybe some further details and training examples.

File: actors/dagger.py [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+        policy_network_state_dict = torch.load(model_data_f)
+        model.policy_network.load_state_dict(policy_network_state_dict)
+
+        # Load version data
+        model.epoch_idx = version_user_data["epoch_idx"]
+        model.total_samples = version_user_data["total_samples"]
+        return model
+
+
+class DaggerStudent:
```
Comment (cloderic) [inline actors/dagger.py:101]:
```suggestion
class DaggerStudentActor:
```

Would be more inline with the other naming

File: actors/dagger.py [hunk 3]
```diff
@@ -0,0 +1,356 @@
...
+
+
+class DaggerTraining:
+    default_cfg = {
+        "seed": 12,
+        "num_data_gather_trials": 50,
+        "num_imitation_trials": 50,
+        "num_mlp_steps": 10,
+        "num_epochs": 4,
+        "num_parallel_trials": 1,
```
Comment (cloderic) [inline actors/dagger.py:136]: Can it support more ? If so make it the default.

File: actors/dagger.py [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+                    # Feed the student's observation to the teacher model to find the correct action
+                    probs = torch.softmax(teacher_model.actor_network(student_observation), dim=-1)
+                    discrete_action_tensor = torch.distributions.Categorical(probs).sample()
+                    action_value = SpaceValue(
+                        properties=[SpaceValue.PropertyValue(discrete=discrete_action_tensor.item())]
+                    )
+                    correct_action = torch.tensor(
+                        flatten(self._environment_specs.action_space, action_value), dtype=self._dtype
+                    )
+                    actions.append(correct_action)
```
Comment (cloderic) [inline actors/dagger.py:318]: Would be easier to follow if this one is named "corrected_actions" or "teacher_actions"

File: actors/dagger.py [hunk 5]
```diff
@@ -0,0 +1,356 @@
...
+
+                run_session.log_metrics(total_reward=sum(r.item() for r in student_rewards))
+                observations.extend(student_observations)
+
+                if len(observations) < self._cfg.batch_size:
+                    continue
+
+                for _ in range(self._cfg.num_mlp_steps):
+                    # Sample a batch of observations/actions
+                    batch_indices = np.random.default_rng().integers(0, len(observations), self._cfg.batch_size)
```
Comment (cloderic) [inline actors/dagger.py:328]: you should create the rng at the beginning (and use the configured seed).

File: actors/dagger.py (hunk 4 above)
Comment (cloderic) [reply actors/dagger.py:318]: you might also want to use/adapt https://github.com/cogment/cogment-verse/blob/next/cogment_verse/replay_buffers/torch_replay_buffer.py

Comment (vabdollahi) [reply actors/dagger.py]: Added a linear schedule similar to the simple_dqn epsilon. I think I'll skip using the replay buffer since I don't need all that info to be stored.
